src/lib/IDBDAO.js

```diff
@@ -0,0 +1,350 @@
+          self.withStore_(mode, function(store) {
+            // console.log('q length: ', q.length);
+            if ( self.q_ == q ) self.q_ = undefined;
+            for ( var i = 0 ; i < q.length ; i++ ) q[i](store);
+          });
+        })();
+      } else {
+        this.q_.push(fn);
+        // Diminishing returns after 10000 per batch
+        if ( this.q_.length == 10000 ) this.q_ = undefined;
+      }
+    },
+
+    function withStore_(mode, fn) {
+      var self = this;
+      if ( self.__TXN__[0] ) {
+        try {
+          fn.call(self, self.__TXN__[0]);
+          return;
+        } catch (x) {
+          self.__TXN__[0] = undefined;
+        }
+      }
+      self.withDB.then(function (db) {
+        var tx = db.transaction([self.name], mode);
+        var os = tx.objectStore(self.name);
+        self.__TXN__[0] = os;
+        fn.call(self, os);
+      });
+    },
+
+    function put(value) {
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.withStore("readwrite", function(store) {
+          var request = store.put(self.serialize(value), value.id);
+          request.transaction.addEventListener(
+            'complete',
+            function(e) {
+              self.pub('on','put', value);
+              resolve(value);
+            });
+          request.transaction.addEventListener(
+            'error',
+            function(e) {
+              reject(self.IDBInternalException.create({ id: value.id, error: e }));
+            });
+        });
+      });
+    },
+
+    function find(key) {
+      var self = this;
+
+      return new Promise(function(resolve, reject) {
+        self.withStore("readonly", function(store) {
+          var request = store.get(key);
+          request.transaction.addEventListener(
+            'complete',
+            function() {
+              if (!request.result) {
+                reject(self.ObjectNotFoundException.create({ id: key }));
+                return;
+              }
+              var result = self.deserialize(request.result);
+              resolve(result);
+            });
+          request.onerror = function(e) {
+            reject(self.IDBInternalException.create({ id: key, error: e }));
+          };
+        });
+      });
+    },
+
+    function remove(obj) {
+      var self = this;
+      var key = obj.id != undefined ? obj.id : obj;
+      return new Promise(function(resolve, reject) {
+        self.withStore("readwrite", function(store) {
+          var getRequest = store.get(key);
+          getRequest.onsuccess = function(e) {
+            if (!getRequest.result) {
+              // not found? as good as removed!
+              self.pub('on','remove', data);
+              resolve();
+              return;
+            }
+            var data = self.deserialize(getRequest.result);
+            var delRequest = store.delete(key);
+            delRequest.transaction.addEventListener('complete', function(e) {
```

Comment:
What happens if we call delete an the object isn't there?  Can we detect that case and treat it as a success.  Then we can skip the entire getRequest step.